docs/source/integrations/building-integrations.md

```diff
@@ -1,54 +1,363 @@
+parse the request body into a JavaScript object. We know that if there's no
+`body` on the request, the middleware isn't configured properly so we respond
+with an error.
+
 
+```ts
+if (!req.body) {
+  res.status(500);
+  res.send(
+    '`req.body` is not set; this probably means you forgot to set up the ' +
+      '`body-parser` middleware before the Apollo Server middleware.',
+  );
+  return;
+}
+```
+
+Because the Express implementation uses the `body-parser` package, all requests
+with a `content-type: application/json` header will either be parsed into a
+JavaScript object or an error will be returned before the handler is even
+called. It's an integration's responsibility to ensure that the body is parsed
+appropriately based on the `content-type` header and that errors are returned
+for invalid requests.
+
+FIXME: add note about qs and get query string parameter parsing
+
+Apollo Server expects a few "types" of requests:
+* `POST` GraphQL Request: `POST` with a `content-type: application/json` header
+  * Request body is a JSON object with the following properties:
+    * `query` (required): GraphQL query string
+    * `variables`: JSON object containing GraphQL variables if provided
+    * `operationName`: GraphQL operation name string if provided
+    * `extensions`: JSON object containing arbitrary extension data if provided
+* `POST` APQ GraphQL Request: `POST` with a `content-type: application/json`
+  header
+  * Request body is a JSON object with the following properties:
+    * `extensions`: JSON object containing a `persistedQuery` object with
+      `version` and `sha256Hash` properties
+* `GET` GraphQL Request: `GET` with a `content-type: text/plain` header
+  * The following query parameters are URL encoded:
+    * `query` (required): GraphQL query string
+    * `variables`: JSON object containing GraphQL variables if provided
+    * `operationName`: GraphQL operation name string if provided
+    * `extensions`: JSON object containing arbitrary extension data if provided
+* `GET` APQ GraphQL Request: `GET` with a `content-type: text/plain` header
+  * The following query parameters are URL encoded:
+    * `extensions`: JSON object containing a `persistedQuery` object with
+      `version` and `sha256Hash` properties
+* Landing page request: `GET` with an `accept: text/html` header
+
+FIXME: restructure this^
+
+#### Constructing the `HTTPGraphQLRequest` object
+
+```ts
+interface HTTPGraphQLRequest {
+  method: string;
+  headers: Map<string, string>;
+  searchParams: any;
+  body: any;
+}
+```
+
+With the request body parsed, we can now construct an `HTTPGraphQLRequest`.
+Apollo Server handles the logic of `GET` vs `POST`, applicable headers, and
+whether to look in `searchParams` or `body` for the GraphQL-specific parts of
+the query.
+
+Express handles the body parsing as well as parsing any query parameters, so
+constructing the `HTTPGraphQLRequest` only requires us to transform the
+`headers` into a `Map` like so:
```

Comment:
calling out that header names need to be lower-case and how to handle duplicate headers in the text does seem reasonable.

Reply:
We leave that up to express, I know we expect only lower cased headers (so we don't worry about casing in express implementation), do we just need to call out how express handles this? i.e. merge duplicate headers with `, ` (i'm guessing, need to look)

Reply:
My point (maybe you've addressed that) is that integration authors are required to ensure all headers put in the HTTPGraphQLRequest.headers Map have lower-case names. In the case of Express this is a no-op but it might not be for others.